packages/origin-247-certificate/src/certificateForUnitTests.service.ts

```diff
@@ -0,0 +1,120 @@
+import { CertificateService } from './certificate.service';
+import { EventBus } from '@nestjs/cqrs';
+import {
+    ICertificate,
+    IClaimCommand,
+    ITransferCommand,
+    ISuccessResponse,
+    IIssuedCertificate,
+    IIssueCommandParams
+} from './types';
+import { Injectable } from '@nestjs/common';
+import { CertificatePersistedEvent } from './externals';
+
+type PublicPart<T> = { [K in keyof T]: T[K] };
+
+@Injectable()
+export class CertificateForUnitTestsService<T> implements PublicPart<CertificateService<T>> {
+    private serial = 0;
+    private db: ICertificate<T>[] = [];
+
+    constructor(private readonly eventBus: EventBus) {}
+
+    public async getAll(): Promise<ICertificate<T>[]> {
+        return [...this.db];
+    }
+
+    public async getById(id: number): Promise<ICertificate<T> | null> {
+        return this.db.find((c) => c.id === id) ?? null;
+    }
+
+    public async issue(params: IIssueCommandParams<T>): Promise<IIssuedCertificate<T>> {
+        this.serial += 1;
+
+        const certificate: ICertificate<T> = {
+            claims: [],
+            claimers: {},
+            creationBlockHash: '',
+            creationTime: Math.floor(Date.now() / 1000),
+            deviceId: params.deviceId,
+            generationStartTime: Math.floor(params.toTime.getTime() / 1000),
+            generationEndTime: Math.floor(params.toTime.getTime() / 1000),
+            id: this.serial,
+            tokenId: this.serial,
+            issuedPrivately: false,
+            metadata: params.metadata,
+            owners: {
+                [params.toAddress]: params.energyValue
+            }
+        };
+
+        this.db.push(certificate);
+
+        setTimeout(() => {
+            this.eventBus.publish(new CertificatePersistedEvent(certificate.id));
+        }, 2000);
+
+        return {
+            ...certificate,
+            issuedPrivately: false,
+            isClaimed: false,
+            isOwned: true,
+            myClaims: [],
+            energy: {
+                claimedVolume: '0',
+                privateVolume: '0',
+                publicVolume: params.energyValue
+            }
+        };
+    }
+
+    public async claim(command: IClaimCommand): Promise<ISuccessResponse> {
+        const certificate = this.db.find((c) => c.id === command.certificateId);
+
+        if (!certificate) {
+            return {
+                success: false,
+                message: `No certificate of ${command.certificateId} found`
+            };
+        }
+
+        certificate.claims.push({
+            claimData: command.claimData,
+            value:
+                Number(command.energyValue) ??
+                Object.values(certificate.owners).reduce((sum, v) => (sum += Number(v)), 0),
```

Comment:
Won't these be BigNumbers?

Reply:
Nope, these are numbers. it's bug probably in issuer-api, that certificate returned has numbers, not strings.

Reply:
Weird, it should return a string.

https://github.com/energywebfoundation/origin/blob/0cdd344dbc718d5048457c95d4d2dada4d45ad96/packages/traceability/issuer-api/src/pods/certificate/certificate.entity.ts#L39

https://github.com/energywebfoundation/origin/blob/37484e58d0d3ac5b4d08612b81c320c70a4f4b49/packages/traceability/issuer/src/blockchain-facade/CertificateUtils.ts#L15

I'll look into it.

Reply:
Where is the data fetched from the Issuer API? This looks like a mocked in-memory db.

Reply:
This is mocked in-memory db.

And by the way, it's IClaim interface

Reply:
Update to the latest canary, this has been fixed. IClaim value is now a string.

Reply:
https://github.com/energywebfoundation/origin/blob/d3b9f857f3119dc89d8a915b6ee96b54137e980a/packages/traceability/issuer/src/blockchain-facade/Certificate.ts#L45

Reply:
fixed